filesystem/filesystem.go

```diff
@@ -496,10 +502,75 @@ func (m *Mount) addMetadata(path string, md metadata.Metadata, owner *user.User)
 	return m.writeDataAtomic(path, data, owner)
 }
 
+func readFull(fd int, data []byte) (int, error) {
+	i := 0
+	for i < len(data) {
+		count, err := unix.Read(fd, data[i:])
+		if err != nil {
+			return 0, err
+		}
+		if count == 0 {
+			break
+		}
+		i += count
+	}
+	return i, nil
+}
+
+// readMetadataFileSafe gets the contents of a metadata file extra-carefully,
+// considering that it could be a malicious file created to cause a
+// denial-of-service.  Specifically, the following checks are done:
+//
+// - It must be a regular file, not another type of file like a symlink or FIFO.
```

Comment:
Is there a reason a symlink can cause denial of service? I can understand it pointing to a weird file, but  we should still be able to _open_ the file. If we then don't set `O_NOFOLLOW`, we would then just check the type of the underlying file.

Also, I thought we wanted to support `.fscrypt` being a symlink? Or does `O_NOFOLLOW` only affect things if the _file_ is a symlink.

Reply:
`O_NOFOLLOW` only affects the final path component.

It's true that the file could always be a weird file directly, so symlinks don't make things *that* much worse.  But symlinks can be created by non-root users pointing to anywhere, whereas creating really weird files like device nodes requires root access.  So it seems desirable to disallow symlinks in this context, given that there is no use case for them here.

Reply:
Updated the comment to explain the above.

Reply:
Sounds reasonable.